File: nim/pack/pack.nim
import strutils, system

type
    VarSpec = tuple
        data: TaintedString # Yes we shouldn't technically require a tainted string. but this works for now.
        namePrefix: string
        fromMemoryFuncName: string
        fromMemoryFileType: string
        fromMemoryReturnType: string

proc appendVar(output: var string, spec: VarSpec) =
    let arrayName = spec.namePrefix & "Array"
    add(output, """

var """ & arrayName & """: array[""")

    let dataLen = len(spec.data)

    add(output, $dataLen)
    add(output, ", char] = [")

    let startLen = len(output)

    for c in spec.data:
        addSep(output, startLen = startLen)
        add(output, "'\\x")
        add(output, toHex(uint8(c)))
        add(output, "'")


    add(output, "]\n")

    let bytesName = spec.namePrefix & "Bytes"

    add(output, """

var """ & bytesName & """: ptr char = cast[ptr char](addr(""" & arrayName & """))

var """ & spec.namePrefix & """*: """ & spec.fromMemoryReturnType & """ = """ & spec.fromMemoryFuncName & """(
    """" & spec.fromMemoryFileType & """",
    """ & bytesName & """,
    """)
    add(output, $dataLen)
    add(output, "\n)\n")


let image: string = readFile("../spritesheet.png")

var spritesheetOutput = """
import raylib
"""

appendVar(spritesheetOutput, (
    data: image,
    namePrefix: "image",
    fromMemoryFuncName: "LoadImageFromMemory",
    fromMemoryFileType: "png",
    fromMemoryReturnType: "Image"
))

writeFile("../spritesheet.nim", spritesheetOutput)

var soundOutput = """
import raylib
"""

appendVar(soundOutput, (
    data: readFile("../sounds/hit1.wav"),
    namePrefix: "hit1",
    fromMemoryFuncName: "LoadWaveFromMemory",
    fromMemoryFileType: "wav",
    fromMemoryReturnType: "Wave"
))

appendVar(soundOutput, (
    data: readFile("../sounds/hit2.wav"),
    namePrefix: "hit2",
    fromMemoryFuncName: "LoadWaveFromMemory",
    fromMemoryFileType: "wav",
    fromMemoryReturnType: "Wave"
))

appendVar(soundOutput, (
    data: readFile("../sounds/treasure.wav"),
    namePrefix: "treasure",
    fromMemoryFuncName: "LoadWaveFromMemory",
    fromMemoryFileType: "wav",
    fromMemoryReturnType: "Wave"
))

appendVar(soundOutput, (
    data: readFile("../sounds/newLevel.wav"),
    namePrefix: "newLevel",
    fromMemoryFuncName: "LoadWaveFromMemory",
    fromMemoryFileType: "wav",
    fromMemoryReturnType: "Wave"
))

appendVar(soundOutput, (
    data: readFile("../sounds/spell.wav"),
    namePrefix: "spell",
    fromMemoryFuncName: "LoadWaveFromMemory",
    fromMemoryFileType: "wav",
    fromMemoryReturnType: "Wave"
))

writeFile("../sound.nim", soundOutput)

File: c/pack/pack.c
#include "stdio.h"

int append_var(FILE* output, char* file_name, char* prefix) {
    FILE* input = fopen(file_name, "rb");

    if (!input) {
        return 2;
    }

    if (fseek(input, 0, SEEK_END) != 0) {
        return 3;
    }

    long input_length = ftell(input);
    if (input_length < 0) {
        return 4;
    }

    fseek(input, 0, SEEK_SET);

    fprintf(output, "static const unsigned char ");
    fprintf(output, "%s", prefix);
    fprintf(output, "_BYTES[%ld] = {", input_length);

    char* sep = "";
    int c; // `int`, not char, is required to handle EOF
    while ((c = fgetc(input)) != EOF) {
        fprintf(output, "%s%u", sep, c);
        sep = ",";
    }

    fprintf(output, "};\n");
    fprintf(output, "static const int ");
    fprintf(output, "%s", prefix);
    fprintf(output, "_LENGTH = %ld;\n", input_length);

    return 0;
}

int main(void) {
    FILE* output = fopen("../asset_bytes.c", "wb");

    if (!output) {
        return 1;
    }

    int error;

#define APPEND(file_name, prefix) \
    error = append_var( \
        output, \
        file_name, \
        prefix \
    ); \
\
    if (error) {\
        return error;\
    }\
\

    APPEND("../assets/spritesheet.png", "SPRITESHEET")
    APPEND("../assets/hit1.wav", "HIT_1")
    APPEND("../assets/hit2.wav", "HIT_2")
    APPEND("../assets/newLevel.wav", "NEW_LEVEL")
    APPEND("../assets/spell.wav", "SPELL")
    APPEND("../assets/treasure.wav", "TREASURE")

    return 0;
}
